components/builder-core/src/metrics.rs

```diff
@@ -24,44 +25,57 @@ pub const APP_NAME: &'static str = "bldr";
 // Statsd Listener Address
 pub const STATS_ENV: &'static str = "HAB_STATS_ADDR";
 
-// Supported metrics
-#[derive(Debug, Clone)]
-pub enum Counter {
-    GithubApi,
-    GithubAuthenticate,
-    GithubEvent,
-    GithubInstallationToken,
-    SearchPackages,
+pub type ApiEndpoint = &'static str;
+pub type InstallationId = u32;
+
+// Public Interface
+////////////////////////////////////////////////////////////////////////
+
+/// Metric identifiers will usually be static `str`s, but some may
+/// need to be dynamically-generated `String`s. With a `Cow`, we can
+/// accept either.
+pub type MetricId = Cow<'static, str>;
+
+/// All metrics must implement the Metric trait, as well as one of the
+/// type marker traits (e.g., `CounterMetric`).
+pub trait Metric {
+    /// Generate the metric name to be used
+    fn id(&self) -> MetricId;
 }
 
-// Supported metrics
-#[derive(Debug, Clone)]
-pub enum Gauge {
-    PackageCount,
+/// Marker trait for counter metrics (don't want to accidentally
+/// increment a gauge metric!)
+pub trait CounterMetric {}
```

Comment:
Should CounterMetric also inherit from Counter?  I _think_ that could save having 2 impls later..?

Reply:
There is no core `Counter`... that's just what I was conventionally calling the crate-scoped metric definitions. This trait was just added to prevent passing a future `Gauge` metric to an `incr` function, etc.